src/vsock/src/vhost_user_vsock_thread.rs

```diff
@@ -0,0 +1,552 @@
+                let conn_map_key = ConnMapKey::new(local_port, peer_port);
+                let mut new_vsock_conn = VsockConnection::new_local_init(
+                    unix_stream,
+                    VSOCK_HOST_CID,
+                    local_port,
+                    self.guest_cid,
+                    peer_port,
+                    self.get_epoll_fd(),
+                );
+                new_vsock_conn.rx_queue.enqueue(RxOps::Request);
+                new_vsock_conn.set_peer_port(peer_port);
+
+                // Add connection object into the backend's maps
+                self.thread_backend
+                    .conn_map
+                    .insert(conn_map_key, new_vsock_conn);
+
+                self.thread_backend
+                    .backend_rxq
+                    .push_back(ConnMapKey::new(local_port, peer_port));
+
+                // Re-register the fd to listen for EPOLLIN and EPOLLOUT events
+                Self::epoll_modify(
+                    self.get_epoll_fd(),
+                    fd,
+                    epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
+                )
+                .unwrap();
+            } else {
+                // Previously connected connection
+                let key = self.thread_backend.listener_map.get(&fd).unwrap();
+                let vsock_conn = self.thread_backend.conn_map.get_mut(&key).unwrap();
+
+                if evset == epoll::Events::EPOLLOUT {
+                    // Flush any remaining data from the tx buffer
+                    match vsock_conn.tx_buf.flush_to(&mut vsock_conn.stream) {
+                        Ok(cnt) => {
+                            vsock_conn.fwd_cnt += Wrapping(cnt as u32);
+                            vsock_conn.rx_queue.enqueue(RxOps::CreditUpdate);
+                            self.thread_backend.backend_rxq.push_back(ConnMapKey::new(
+                                vsock_conn.local_port,
+                                vsock_conn.peer_port,
+                            ));
+                        }
+                        Err(e) => {
+                            dbg!("Error: {:?}", e);
+                        }
+                    }
+                    return;
+                }
+
+                // Unregister stream from the epoll, register when connection is
+                // established with the guest
+                Self::epoll_unregister(self.epoll_file.as_raw_fd(), fd).unwrap();
+
+                // Enqueue a read request
+                vsock_conn.rx_queue.enqueue(RxOps::Rw);
+                self.thread_backend
+                    .backend_rxq
+                    .push_back(ConnMapKey::new(vsock_conn.local_port, vsock_conn.peer_port));
+            }
+        }
+    }
+
+    /// Allocate a new local port number.
+    fn allocate_local_port(&mut self) -> Result<u32> {
+        // TODO: Improve space efficiency of this operation
+        // TODO: Reuse the conn_map HashMap
```

Comment:
can't we just use a bind(0) and let the kernel safely allocate a local port for us?

Reply:
If I understand this question correctly, I would say no. This is because we don't use the kernel for emulating the device (or anything else) as it is done in the userspace.